packages/squiggle-lang/src/public/SqValuePath.ts

```diff
@@ -1,133 +1,257 @@
-            ...findLoop(ast.value),
-          ];
-        }
-        case "Block": {
+
           if (
-            ast.statements.length === 1 &&
-            ["Array", "Dict"].includes(ast.statements[0].type)
+            pair.type === "KeyValue" &&
+            pair.key.type === "String" // only string keys are supported
           ) {
-            return findLoop(ast.statements[0]);
+            return [
+              SqValuePathEdge.fromKey(pair.key.value),
+              ...buildRemainingPathEdges(pair.value),
+            ];
+          } else if (pair.type === "Identifier") {
+            return [SqValuePathEdge.fromKey(pair.value)]; // this is a final node, no need to buildRemainingPathEdges recursively
           }
         }
+        return [];
       }
-      return [];
-    };
+      case "Array": {
+        for (let i = 0; i < ast.elements.length; i++) {
+          const element = ast.elements[i];
+          if (locationContains(element.location, offset)) {
+            return [
+              SqValuePathEdge.fromIndex(i),
+              ...buildRemainingPathEdges(element),
+            ];
+          }
+        }
+        return [];
+      }
+      case "LetStatement": {
+        return [
+          SqValuePathEdge.fromKey(ast.variable.value),
+          ...buildRemainingPathEdges(ast.value),
+        ];
+      }
+      case "DefunStatement": {
+        return [
+          SqValuePathEdge.fromKey(ast.variable.value),
+          ...buildRemainingPathEdges(ast.value),
+        ];
+      }
+      case "Block": {
+        if (
+          ast.statements.length === 1 &&
+          ["Array", "Dict"].includes(ast.statements[0].type)
+        ) {
+          return buildRemainingPathEdges(ast.statements[0]);
+        }
+      }
+    }
+    return [];
+  }
+  return buildRemainingPathEdges(ast);
+}
+
+export class SqValuePath {
+  public root: ValuePathRoot;
+  public edges: SqValuePathEdge[];
 
+  constructor(props: { root: ValuePathRoot; edges: SqValuePathEdge[] }) {
+    this.root = props.root;
+    this.edges = props.edges;
+  }
+
+  static findByAstOffset({
+    ast,
+    offset,
+  }: {
+    ast: ASTNode;
+    offset: number;
+  }): SqValuePath | undefined {
     return new SqValuePath({
       root: "bindings", // not important, will probably be removed soon
-      items: findLoop(ast),
+      edges: astOffsetToPathEdges(ast, offset),
+    });
+  }
+
+  isRoot() {
+    return this.edges.length === 0;
+  }
+
+  lastItem(): SqValuePathEdge | undefined {
+    return this.edges[this.edges.length - 1];
+  }
+
+  extend(edge: SqValuePathEdge) {
+    return new SqValuePath({
+      root: this.root,
+      edges: [...this.edges, edge],
     });
   }
 
-  itemsAsValuePaths({ includeRoot = false }) {
+  uid(): string {
+    return `${this.root}--${this.edges.map((edge) => edge.uid()).join("--")}`;
+  }
+
+  // Checks if this SqValuePath completely contains all of the nodes in this other one.
+  contains(smallerItem: SqValuePath) {
```

Comment:
How about ``hasAsPrefix`` or ``hasPrefix``

Reply:
That works too. I suggested `startsWith` because of [String.startsWith](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith).